### DISCUSSION TITLE
Current issues to look for

### DISCUSSION THREAD

**Discussion by BreadDain** [IHateMyKite/UnforgivingDevices#31] [General]
I can't really report them as bugs because I am neither sure if it is not just my install, neither if it is not part of WIP and is temporarily broken, since I have to use latest commit, nor I do not know how to reproduce the bug or how what is it caused by.

~~Hardcore mode: had several times the issue where controls were locked completely and I couldn't do anything. Once it was randomly (maybe after stripping one items) and another time it was already after I disabled the option and mcm and get rid from an armbinder. Something's wrong for sure~~ reported

~~Another Hardcore mode issue: latest commit won't allow the device selection menu to be called.~~ changed to C keyheld

~~Third person switch: code implies that every time one starts a minigame camera should switch to 3rd person... which is never the case for me. I blame my own setup, but this is likely a mod conflict to look for, suspected Better Third Person Selection, True Directional Movement and Smoothcam. Other users are likely to report this in future if anyone notices that with the same setup.~~ removed from zadlibs patch, intended

Followers don't gain orgasm with vibration on latest commit. Maybe it is due to specific conditions or exposure mult settings of the follower, but their arousal is totally not locked, and they were at 100 arousal, and exposure mult was more than 0 value. I assume something wrong with UD itself, should be easy to find

~~Elbow shackles have very strange animation behaviour, or more specifically its specific animations won't start at all. Happened 2 times when I was equipped with them, something wrong with zad libs patch? EDIT: yoke is affected as well EDIT2: valid for beta 11, happens very often right after first game load.~~ DAR interference?

Device unlock timeout, bugs vibrations
![изображение](https://user-images.githubusercontent.com/30041703/166899087-1e328f65-a426-45fa-b4db-1d02cf762f13.png)

Unequipping devices while vibration is triggered/active might produce zombie stack (device is not present on acto)
![изображение](https://user-images.githubusercontent.com/30041703/167042731-e2efed28-ba11-4ac5-8f27-184f56480a23.png)

**Comment by IHateMyKite** [comment]
> Followers don't gain orgasm with vibration on latest commit. Maybe it is due to specific conditions or exposure mult settings of the follower, but their arousal is totally not locked, and they were at 100 arousal, and exposure mult was more than 0 value. I assume something wrong with UD itself, should be easy to find

Can't replicate. Did you use clean save? Also, try to register the NPCs if you didn't already

**Reply by BreadDain** [reply]
> > Followers don't gain orgasm with vibration on latest commit. Maybe it is due to specific conditions or exposure mult settings of the follower, but their arousal is totally not locked, and they were at 100 arousal, and exposure mult was more than 0 value. I assume something wrong with UD itself, should be easy to find
> 
> Can't replicate. Did you use clean save? Also, try to register the NPCs if you didn't already

I did both, I think the issue might be related to Aroused somehow. Started a new save since beta 10 and switched to OSL, will look more into it.

**Comment by IHateMyKite** [comment]
> Device unlock timeout, bugs vibrations
> 
> Unequipping devices while vibration is triggered/active might produce zombie stack (device is not present on acto)

Hopefully fixed

**Comment by BreadDain** [comment]
~~Check possibility of activating clitoral piercings by belt activations. While I was digging in code I think I've missed it, there was plug activations but not piercings.~~

> Function deflatePlug(int decrease)
> 	If UD_DeviceKeyword == libs.zad_DeviousPlugAnal
> 		libs.DeflateAnalPlug(getWearer(), decrease)
> 	Else	
> 		libs.DeflateVaginalPlug(getWearer(), decrease)
> 	EndIf
> 	
> 	_inflateLevel -= 1
> 	if _inflateLevel < 0
> 		_inflateLevel = 0
> 	endif
> 	deflateprogress = 0.0
> EndFunction

~~Seems like inflateLevel always decreases by 1 instead of "decrease" value, intended? It is NPC-specific value, so probably invisible bug in game~~

**Reply by IHateMyKite** [reply]
> Check possibility of activating clitoral piercings by belt activations. While I was digging in code I think I've missed it, there was plug activations but not piercings.

I have only added activation of plugs. Adding piercing will be possible but it will require some more work.

> 
> > Function deflatePlug(int decrease)
> > If UD_DeviceKeyword == libs.zad_DeviousPlugAnal
> > libs.DeflateAnalPlug(getWearer(), decrease)
> > Else
> > libs.DeflateVaginalPlug(getWearer(), decrease)
> > EndIf
> > _inflateLevel -= 1
> > if _inflateLevel < 0
> > _inflateLevel = 0
> > endif
> > deflateprogress = 0.0
> > EndFunction
> 
> Seems like inflateLevel always decreases by 1 instead of "decrease" value, intended? It is NPC-specific value, so probably invisible bug in game

My bad. This is bug. Thanks for finding it. But there are nowhere currently function calls that have decrease other than 1, so it was not causing trouble (for now).

**Comment by IHateMyKite** [comment]
> Elbow shackles have very strange animation behaviour, or more specifically its specific animations won't start at all. Happened 2 times when I was equipped with them, something wrong with zad libs patch? EDIT: yoke is affected as well EDIT2: valid for beta 11, happens very often right after first game load. DAR interference?

Most certainly caused by DAR. DAR update idle animation based on some conditions. Its possible that the idle was changed right after device have used its own bound idle. You can try to fix this by editing DAR mod conditions.txt so it checks that actor doesn't have heavy bondage device which changes idle

**Comment by IHateMyKite** [comment]
> Device unlock timeout, bugs vibrations ![изображение](https://user-images.githubusercontent.com/30041703/166899087-1e328f65-a426-45fa-b4db-1d02cf762f13.png)

Did this happen on NPC or player? Also, is the problem still present?

> Unequipping devices while vibration is triggered/active might produce zombie stack (device is not present on acto) ![изображение](https://user-images.githubusercontent.com/30041703/167042731-e2efed28-ba11-4ac5-8f27-184f56480a23.png)

Hopefully fixed. Added check so the loop will end if render device is no longer equipped

**Comment by BreadDain** [comment]
> Did this happen on NPC or player? Also, is the problem still present?

Likely it is gone, I've reported that part long ago, just used to edit original post which was bad decision.
Most of these reports are just a quick dump of everything strange I can notice without being able to replicate.

> Most certainly caused by DAR. DAR update idle animation based on some conditions. Its possible that the idle was changed right after device have used its own bound idle. You can try to fix this by editing DAR mod conditions.txt so it checks that actor doesn't have heavy bondage device which changes idle

~~The problem is that with vanilla DD that was not the case, it started fairly recently specifically with UD. And only with specific devices somehow.
Fixing DAR conditions is a workaround, but... there is a waaaaay too many of them, so unless I write some sort of autopatcher no way I gonna make it manually. Count is on hundreds if not thousands.~~ Annihilated

**Comment by BreadDain** [comment]
2.0b13
Struggle loop fails at the first tick, resulting in immediate struggle cancelation right from the start. Issue was present in one of previous UD versions. (Edit: not sure if it is not related to my bad update transition)
Big amount of different errors in log, related to different mods, Strike, Cursed Loot, etc. They keep happening and have different sources of issues.
~~Minigame delays are back, hard, if not impossible, to score critical hits. Related not just to cutting minigame, but all of them in general.~~ bad update transition for me
Cutting minigame - spamming action button (hard!, I mean it!) still freezes stamina drain like it was before.
Cuffs have uneven amount of locks, intended?
Patched rope devices appear to be unable to be cut, intended?

**Reply by IHateMyKite** [reply]
> 2.0b13 Struggle loop fails at the first tick, resulting in immediate struggle cancelation right from the start. Issue was present in one of previous UD versions. (Edit: not sure if it is not related to my bad update transition) 

That is strange. This never happened to me. Did you managed to replicate this?

> Big amount of different errors in log, related to different mods, Strike, Cursed Loot, etc. They keep happening and have different sources of issues. 

They are not really issues. They are more of a warning. I have added Warning() and Info() functions, so it can be better categorized. If error is called, it means that there is something seriously wrong. Warning only means, well, warning. Aka, nothings wrong, but it could be done better. Info is everything else. Most devices like Cursed Loot and Devious Strike doesn't really check if they can lock actor in the device they want to. Because of that, the lock function fails, which may or may not result in an issue (depends on what device is being locked on).

> Cutting minigame - spamming action button (hard!, I mean it!) still freezes stamina drain like it was before. Cuffs have uneven amount of locks, intended? Patched rope devices appear to be unable to be cut, intended?

This is mostly intended, I removed the widget update function from key event. Before, every time the special key was press, the widget was updated. But there was problem that the widget update is slow and so it was causing performance issue. Because of that, the widget is now updated in main minigame loop like in other minigames. If you press slow, it will increase by small amount. If you press more it will increase by big amount.

**Reply by BreadDain** [reply]
> That is strange. This never happened to me. Did you managed to replicate this?

Yes, it happens from time to time, not often, usually in the middle of struggling from set of various devices. Something tells me that this is connected to some strange AV/start parameters.

**Reply by BreadDain** [reply]
Was able to catch this during the issue:
![изображение](https://user-images.githubusercontent.com/30041703/178075206-4cb6a43a-cae6-4ca0-ada2-cae1fca7f962.png)

**Comment by BreadDain** [comment]
Checked a save file after an especially laggy key spam minigame, found out this. The minigame has already finished, but the scripts were still executing.
![изображение](https://user-images.githubusercontent.com/30041703/177790486-2da53e18-38d2-4909-813e-714353e33ea1.png)

**Reply by IHateMyKite** [reply]
I think that this is normal. They are active scripts so once you load the game and wait a minute, it should disapear. Otherwise the cause is same as in previous post. Cutting/Fourceout minigame are slow. I already added some kind of limiter, so it is not possible to cause sespended scripts, but I will have to give it a more time. You can give it a look [here](https://github.com/IHateMyKite/UnforgivingDevices/blob/9a9e91b41448e91b61fd741564db13f5cc8de4b8/Scripts/Source/UDCustomDeviceMain.psc#L1345-L1351), maybe you will have some idea on how to improve it. I didn't give the design much thinking. 
Maybe I could make new AME that will be used for this kind of minigame. It will contain just key processing and nothing else. Because it will be used only in minigame there will be no need for some checks that are currently in OnKeyDown.

**Reply by BreadDain** [reply]
Missing [return](https://github.com/IHateMyKite/UnforgivingDevices/blob/9a9e91b41448e91b61fd741564db13f5cc8de4b8/Scripts/Source/UDCustomDeviceMain.psc#L1393)? But it is not much anyway, what I think is to try utilize script [State](https://www.creationkit.com/index.php?title=State_Reference) functionality for different types of minigames, to prevent it checking every possible condition. Also very curious how exactly IsInMenuMode() impacts delay and if offloading it to State-specific onUpdate event with 1 sec frequency wouldn't be a better solution. But all this will take some time to think over and code.
Also trying to ask some people who had some knowledge on hotkeys how exactly Papyrus handles them since there is not much info. It seems people are trying to offload quick functions like this to SKSE, which is a whole another level.

**Comment by BreadDain** [comment]
Something is off with the plug struggle if an NPC (follower) wears Abadon plug during the quest. It takes damage super slowly, much less than usual.

**Reply by IHateMyKite** [reply]
Will investigate.

**Comment by BreadDain** [comment]
MAH modifier sometimes fail to work properly if it attempts to equip 6+ devices, guarantee to fail at 7

**Comment by BreadDain** [comment]
Inability to move a device from stacked amount from one inventory to another (from player to follower), opens interaction menu instead.

In other words, having 2 same devices with intention to equip it on two different actors starting from player will disallow equipping it on follower after, producing mod's interaction UI instead.

**Reply by IHateMyKite** [reply]
I think I understand the issue. Will give it a look

**Comment by BreadDain** [comment]
Found an interesting (buggy) interaction with UD, when a follower has wrist restraint equipped and player pulls a weapon to combat-ready state follower might start trying to equip and pull out a weapon in a quick succession. ~~Possibly Nether's Follower Framework is causing this, likely - AI Package based issue.~~ That's related to unique Serana's brains, she is constantly switching packages and bypasses all calming effects.

**Reply by IHateMyKite** [reply]
This is possible issue with all NPC. I still didn't found way to prevent NPC from drawing weapon. Even if I remove the weapon from inventory, they just magicali create new one.

**Reply by BreadDain** [reply]
I probably know how to deal with this, will test it later on. Idea is overriding it with no-weapons high priority package on effect start and removing override later.

**Comment by BreadDain** [comment]
Struggling (from gag) while wearing a yoke plays armbinder animation.

**Reply by IHateMyKite** [reply]
Is this only relevant for gag or for all devices ?

**Reply by BreadDain** [reply]
All devices

**Reply by IHateMyKite** [reply]
Can't replicate. What yoke are you wearing?

**Reply by IHateMyKite** [reply]
This should already be fixed with  7b41bb. Please check the my comment in issue you added.

**Comment by BreadDain** [comment]
Bad compatibility with DD blindfold effects (other than dark fog) and Hardcore effect, controls restriction misbehaves.

**Reply by IHateMyKite** [reply]
Know about this but I'm scared to tinker with it as it will be pain in the ass. But I guess there no running away now...
